sips/dynamic_qbft_timeout.md

```diff
@@ -0,0 +1,90 @@
+| Author         | Title                 | Category | Status   |
+| -------------- | --------------------- | -------- | -------- |
+| Gal Rogozinski |  QBFT timeout | Core     | approved |
+
+
+**Summary**  
+Changes to the constant timeouts we defined in [SIP#6](https://github.com/bloxapp/SIPs/blob/main/sips/constant_qbft_timeout.md) that should improve performance. 
+
+**Rational**  
+Currently for all duties, we have a `quickTimeout` of 2 seconds before sending a `RoundChange` message and a `slowTimeout` of 2 minutes starting from the 8th round.The timer starts when the QBFT instance starts. This approach has 2 issues:
+
+1. Duties that require a pre-consensus stage may start their instances at different times, causing some nodes to timeout sooner than others. This will result in a greater probability of a failed first round.
+2. Different duties have different timing assumptions, which may cause some duties to execute earlier than others.
+
+
+
+**Beacon Duty Timing Assumptions**  
+The Beacon chain has some timing assumptions regarding duty execution for maximizing rewards.
+
+| Duty           | Timing Assumption | Dependency     | Description                                                                                                         |
+| -------------- | ----------------- | -------------- | ------------------------------------------------------------------------------------------------------------------- |
+| Proposal       | < 4 Sec           | ---            | Attestations wait 4 seconds for a block                                                                             |
+| Attester       | < 4 Sec           | Proposal       | Attestations wait 4 seconds for a block then need to be propogated within 4 seconds for aggregators to pick them up |
+| Aggregator     | < 4 Sec           | Attester       | Aggregators start aggregating within 8 seconds from slot start and need to propogate within 4 seconds               |
+| Sync Committee | < 4 sec           | Proposal       | Sync Committee wait 4 seconds for a block                                                                           |
+| Contribution   | < 4 sec           | Sync Committee | Contributions wait 8 seconds from the slot start                                                                    |
+
+        
+
+**Specification**
+
+From the slot start time after `baseDuration` delay wait 2 seconds before emitting a `RoundChange` message until the 6th round (inclusive). Afterwards, wait 2 minutes.
+
+```go
+var (
+	// quickTimeoutThreshold is the round after which the timeout duration increases to 2 minutes
+	quickTimeoutThreshold = Round(6)
+	// quickTimeout is the timeout in seconds for the first 6 rounds
+	quickTimeout int64 = 2 // 2 seconds
```

Comment:
IMO `time.Duration` is more readable

```suggestion
	quickTimeout time.Duration = 2 * time.Second
```

Reply:
I was trying to move away from using go libs in SIPs
The idea is that the spec shouldn't be for helping implement a node in any language

Also for future SIPs I want to start using python like pseudo-code

Reply:
Just my 2 cents,

I would say while we still use Golang might as well do it as Golang prefers to do it (as Nikita mentioned), but the idea to use pseudo-code is the right one ofc

the even more important thing though is that spec "provides interface and not implementation" - I elaborate a bit on it [here](https://github.com/ssvlabs/SIPs/pull/37#issuecomment-2445303530)

Reply:
> as long as we keep away from using advanced Python-specific syntax
That's why I said "python like pseudo code" :-) 

> spec "provides interface and not implementation" 
Currently spec has too many implementation details including full fledged consensus impl.
This is due to legacy reasons and how the project was inherited. We are planning a Markdown spec.
All implementations that we will have will just be used for creating tests.